### ISSUE TITLE
GRPO训练internVL3时，使用vllm出现format reward始终为0，但不使用vllm时reward会正常上升

### ISSUE THREAD

**Issue by SongCASIA** [modelscope/ms-swift#5271]
使用vllm训练：
```
# 8*32G GPU
nproc_per_node=8

CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=$nproc_per_node \
swift rlhf \
    --rlhf_type grpo \
    --model internvl3-1b/ \
    --model_type internvl3 \
    --reward_funcs format \
    --train_type lora \
    --lora_rank 8 \
    --lora_alpha 32 \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.5 \
    --vllm_tensor_parallel_size 2 \
    --vllm_limit_mm_per_prompt '{"image": 8}' \
    --dataset object_rel_direction_medium_rlhf.jsonl \
    --torch_dtype float16 \
    --num_train_epochs 1 \
    --max_length 8192 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --eval_steps 100 \
    --save_steps 100 \
    --learning_rate 1e-5 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --output_dir output_8gpu \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --max_completion_length 2048 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --system examples/train/grpo/prompt.txt \
    --deepspeed zero2 \
    --temperature 1.0 \
    --top_p 0.99 \
    --top_k 50 \
    --log_completions true \
    --async_generate false \
    --move_model_batches 16 \
    --offload_optimizer true \
    --offload_model true \
    --gc_collect_after_offload true \
    --sleep_level 1
```
不使用vllm训练：
```
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model internvl3-1b/ \
    --model_type internvl3 \
    --reward_funcs format \
    --train_type lora \
    --lora_rank 8 \
    --lora_alpha 32 \
    --dataset object_rel_direction_medium_rlhf.jsonl \
    --torch_dtype float16 \
    --max_completion_length 2048 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --learning_rate 1e-5 \
    --gradient_accumulation_steps 1 \
    --eval_steps 100 \
    --save_steps 100 \
    --save_total_limit 2 \
    --logging_steps 5 \
    --max_length 8192 \
    --output_dir output \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --temperature 1. \
    --top_p 0.99 \
    --top_k 50 \
    --system 'examples/train/grpo/prompt.txt' \
    --deepspeed zero2 \
    --log_completions true
```
使用vllm时，format reward的mean、std均为0，但不使用vllm时format reward会稳定上升至0.9+，请问这是什么问题导致的？

**Comment by hjh0119** [comment]
There may be some discrepancies between vllm and PT during generation. 

You could try running inference directly with both PT and vllm to compare their outputs.

For further debugging, you can set `report_to wandb` or `report_to swanlab` to visually monitor the model’s generated responses.

**Comment by SongCASIA** [comment]
> There may be some discrepancies between vllm and PT during generation.
> 
> You could try running inference directly with both PT and vllm to compare their outputs.
> 
> For further debugging, you can set `report_to wandb` or `report_to swanlab` to visually monitor the model’s generated responses.

非常感谢您的回复。由于之前的infer是使用pt进行的，因此花了一些时间在vllm上验证。
我们在infer的过程中发现pt与vllm存在很多差异，例如，使用vllm进行infer时，会存在token长度的问题：
```
WARNING 08-06 02:38:57 [model_runner.py:1305] Computed max_num_seqs (min(256, 30000 // 106496)) to be less than 1. Setting it to the minimum value of 1.
Token indices sequence length is longer than the specified maximum sequence length for this model (106560 > 12288). Running this sequence through the model will result in indexing errors
WARNING 08-06 02:38:59 [profiling.py:245] The sequence length used for profiling (max_num_batched_tokens / max_num_seqs = 30000) is too short to hold the multi-modal embeddings in the worst case (106560 tokens in total, out of which {'image': 106496} are reserved for multi-modal embeddings). This may cause certain multi-modal inputs to fail during inference, even when the input text is short. To avoid this, you should increase `max_model_len`, reduce `max_num_seqs`, and/or reduce `mm_counts`.
```
但是使用pt则不会出现这种问题，只需要简单的启动指令即可：
```
CUDA_VISIBLE_DEVICES=0 VIDEO_SEGMENTS=32 \
	swift infer \
		--model internvl3-1b/ \
		--model_type internvl3  \
		--stream true \
		--infer_backend pt \
		--temperature 0.0  \
		--max_new_tokens 2048 \
		--val_dataset dataset/object_rel_direction_medium.jsonl \
		--max_model_len 8192 \
		--result_path result_json/object_rel_direction_medium_finetune_pt.jsonl
```
而vllm则需要增设诸多关于token和model长度的参数。
观察infer结果也可以发现，二者的生成结果存在一定的差异。
综合上述vllm与pt在生成阶段的差异表现，我认为这可能是导致使用vllm训练grpo时reward保持在0的原因，那么请问该如何缓解这种问题？或者我需要查看和修改哪一部分代码来进行符合任务的设置？

**Comment by hjh0119** [comment]
Based on experience, there is usually a difference between PT and vllm inference, but the gap should not be very large. Therefore, I suggest you check whether there is something wrong with your vllm inference settings.

For example, {'image': 106496} looks unusual.

**Comment by SongCASIA** [comment]
这里确实有问题，但实际上，使用vllm进行infer时加入了`--limit_mm_per_prompt`参数：
```
CUDA_VISIBLE_DEVICES=0 VIDEO_SEGMENTS=32 \
	swift infer \
		--model internvl3-1b/ \
		--model_type internvl3  \
		--stream true \
		--torch_dtype float16 \
		--infer_backend vllm \
		--limit_mm_per_prompt '{"image": 32}' \
		--temperature 0.0  \
		--max_new_tokens 2048 \
		--val_dataset dataset/object_rel_direction_medium.jsonl \
		--max_model_len 8192 \
		--result_path result_json/object_rel_direction_medium_finetune_vllm.jsonl
```
加入该参数之后仍然出现了`{'image': 106496}`

**Comment by SongCASIA** [comment]
> Based on experience, there is usually a difference between PT and vllm inference, but the gap should not be very large. Therefore, I suggest you check whether there is something wrong with your vllm inference settings.
> 
> For example, {'image': 106496} looks unusual.

我分别使用pt和vllm进行评测，指令分别如下：
```
CUDA_VISIBLE_DEVICES=0 VIDEO_SEGMENTS=8 \
	swift infer \
		--model internvl3-1b/ \
		--model_type internvl3  \
		--stream true \
		--torch_dtype float16 \
		--infer_backend vllm \
		--limit_mm_per_prompt '{"image": 8}' \
		--temperature 0.0  \
		--max_new_tokens 2048 \
		--val_dataset dataset/object_rel_direction_medium.jsonl \
		--result_path result_json/object_rel_direction_medium_finetune_vllm.jsonl

CUDA_VISIBLE_DEVICES=0 VIDEO_SEGMENTS=32 \
	swift infer \
		--model internvl3-1b/ \
		--model_type internvl3  \
		--stream true \
		--infer_backend pt \
		--temperature 0.0  \
		--max_new_tokens 2048 \
		--val_dataset dataset/object_rel_direction_medium.jsonl \
		--max_model_len 8192 \
                --gpu_memory_utilization 0.5
		--result_path result_json/object_rel_direction_medium_finetune_pt.jsonl
```
对比response的结果可以发现，最终评测结果的差异性并不大：
```
pt:
{"response": "A", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "A", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "C", 
{"response": "A", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B", 
{"response": "A", "labels": "C", 
{"response": "B", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "A",
{"response": "B", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "A", "labels": "C", 
{"response": "B", "labels": "B", 
{"response": "A", "labels": "A", 
{"response": "B", "labels": "B", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B",
```
```
vllm:
{"response": "C", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "A", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "C",
{"response": "C", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "B", 
{"response": "A", "labels": "A", 
{"response": "C", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "A", "labels": "C", 
{"response": "B", "labels": "B", 
{"response": "B", "labels": "A", 
{"response": "B", "labels": "B", 
{"response": "C", "labels": "C", 
{"response": "B", "labels": "C", 
{"response": "B", "labels": "A", 
{"response": "C", "labels": "B", 
```
当然这是对于选择题的生成结果，是不是可以理解为：当训练GRPO时需要生成<think>...<\think>，completions的token长度越大，生成结果的差异越大，因此导致了使用pt时format reward能够稳定上升至0.9+，但使用vllm时reward始终为0的现象。

**Comment by hjh0119** [comment]
I suspect the difference is more likely related to length related arguments, which might cause vllm to generate outputs without proper endings, such as missing  `<\answer>` or  `<\think>` tags. This would result in the format reward being zero. Therefore, I suggest you check the completions generated by vllm during rollout.

> For further debugging, you can set report_to wandb or report_to swanlab to visually monitor the model’s generated responses.

**Comment by SongCASIA** [comment]
> I suspect the difference is more likely related to length related arguments, which might cause vllm to generate outputs without proper endings, such as missing `<\answer>` or `<\think>` tags. This would result in the format reward being zero. Therefore, I suggest you check the completions generated by vllm during rollout.


我查看了vllm在训练grpo时的生成结果，确实几乎没有生成</think>字段。
推测可能是由于`--max_completion_length`参数对于vllm和pt的限制方式是不同的？因为在训练时这些长度设置实际上是一致的。

**Comment by hjh0119** [comment]
Maybe you can check the difference in generated sequence lengths during training.

**Comment by SongCASIA** [comment]
> Maybe you can check the difference in generated sequence lengths during training.

通过对比vllm和pt生成的completions长度，可以发现vllm生成的长度远大于pt。
观察pt的生成结果，还可以发现生成的长度越长，越难以生成有效的format。
所以vllm倾向于生成更长的</think>过程？以至于grpo训练时难以学习format？
或者这个问题仅针对于video这一模态的训练过程？

**Comment by hjh0119** [comment]
You can submit an issue to vllm

**Comment by SongCASIA** [comment]
vllm与HF/transformers的代码差异性极大，尽管整体代码实现没有太大出入，但部分中间变量的权重仍然有差别，因此难以定位具体原因，所以只能通过调参的方法尽可能缓解vllm出现的问题

**Comment by Anis-W** [comment]
Hi， 请问可以提供一下你的环境信息吗？我在对自行指令微调过的internVL3-1B-Instruct进行GRPO训练时也遇到了使用vllm做采样时生成长度非常长的问题，并且模型还会重复输出 <|endoftext|> 直到 max_completion_length 上限。
